osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
     /// </summary>
     public class Aim : OsuStrainSkill
     {
-        private const double angle_bonus_begin = Math.PI / 3;
-        private const double timing_threshold = 107;
-
         public Aim(Mod[] mods)
             : base(mods)
         {
         }
 
-        protected override double SkillMultiplier => 26.25;
+        protected override int HistoryLength => 2;
+
+        protected override double SkillMultiplier => 24.75;
         protected override double StrainDecayBase => 0.15;
 
+        private const double wide_angle_multiplier = 1.0;
+        private const double acute_angle_multiplier = 1.0;
+        private const double rhythm_variance_multiplier = 1.0;
+
         protected override double StrainValueOf(DifficultyHitObject current)
         {
-            if (current.BaseObject is Spinner)
+            if (current.BaseObject is Spinner || Previous.Count <= 1)
                 return 0;
 
-            var osuCurrent = (OsuDifficultyHitObject)current;
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+            var osuPrevObj = (OsuDifficultyHitObject)Previous[0];
+            var osuLastObj = (OsuDifficultyHitObject)Previous[1];
 
-            double result = 0;
+            var currVector = Vector2.Divide(osuCurrObj.JumpVector, (float)osuCurrObj.StrainTime);
+            var prevVector = Vector2.Divide(osuPrevObj.JumpVector, (float)osuPrevObj.StrainTime);
 
-            if (Previous.Count > 0)
-            {
-                var osuPrevious = (OsuDifficultyHitObject)Previous[0];
+            // Start with regular velocity.
+            double aimStrain = currVector.Length;
 
-                if (osuCurrent.Angle != null && osuCurrent.Angle.Value > angle_bonus_begin)
+            if (Precision.AlmostEquals(osuCurrObj.StrainTime, osuPrevObj.StrainTime, 10)) // Rhythms are the same.
+            {
+                if (osuCurrObj.Angle != null)
                 {
-                    const double scale = 90;
+                    double angle = osuCurrObj.Angle.Value;
+
+                    // Rewarding angles, take the smaller velocity as base.
+                    double angleBonus = Math.Min(currVector.Length, prevVector.Length);
 
-                    var angleBonus = Math.Sqrt(
-                        Math.Max(osuPrevious.JumpDistance - scale, 0)
-                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
-                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
-                    result = 1.4 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
+                    double wideAngleBonus = calcWideAngleBonus(angle);
+                    double acuteAngleBonus = calcAcuteAngleBonus(angle);
+
+                    if (osuCurrObj.StrainTime > 100)
```

Comment:
This should probably state somewhere how fast 100 straintime is

Reply:
done.